### PR TITLE
Improve the performance of Stride when Base isa Collection

### PR DESCRIPTION
**Pull request by ollieatkinson**
Implement StrideSequence so that the default implementation for collection uses IndexingIterator for faster hops between each element.

Closes https://github.com/apple/swift-algorithms/issues/63

Thank you to @timvermeulen for raising the issue, and thanks to @natecook1000 for being awesome! 

**Proof**

(**old**) Stride.Iterator:
```
average: 0.113, relative standard deviation: 2.817%, values: [0.117450, 0.111180, 0.111465, 0.110857, 0.116854, 0.117716, 0.111387, 0.116178, 0.111040, 0.108650]
```

(**new**) Using IndexingIterator:
```
average: 0.000, relative standard deviation: 79.629%, values: [0.000903, 0.000230, 0.000222, 0.000192, 0.000186, 0.000198, 0.000183, 0.000183, 0.000188, 0.000186]
```

-

```swift
func testMeasureStride() {
  measure {
    let s: Stride = (0..<100_000).striding(by: 1_000)
    for _ in s { }
  }
}
```

and to prove the different implementations have the correct performance characteristics the following tests can be performed and output validated

```swift
func testMeasureSequence() {
  measure {
    let s: StrideSequence = (0..<100_000).striding(by: 1_000)
    for _ in s { }
  }
}
  
func testMeasureCollection() {
  measure {
    let c: Stride = (0..<100_000).striding(by: 1_000)
    for _ in c { }
  }
}
```

### Checklist
- [x] I've added at least one test that validates that my change is working, if appropriate (included comments, was unsure if I should include a measure as there are no-other performance tests)
- [x] I've followed the code style of the rest of the project
- [x] I've read the [Contribution Guidelines](../CONTRIBUTING.md)

### CODE REVIEW

File: Sources/Algorithms/Stride.swift [hunk 1]
```diff
@@ -88,12 +102,36 @@ extension Stride: Sequence {
...
+}
+
+extension Stride {
+  @inlinable
+  public func striding(by step: Int) -> Self {
+    Stride(base: base, stride: stride * step)
+  }
+}
+
+extension Stride: Sequence & Collection {
```
Comment (xwu) [inline Sources/Algorithms/Stride.swift:133]:
```suggestion
extension Stride: Sequence, Collection {
```

Nit: This is the more idiomatic spelling.

Comment (timvermeulen) [reply Sources/Algorithms/Stride.swift:133]: Or simply `extension Stride: Collection`, as this also implies `Sequence` 🙂

File: Sources/Algorithms/Stride.swift [hunk 2]
```diff
@@ -33,8 +53,8 @@ extension Sequence {
   }
 }
 
-/// A wrapper that strides over a base sequence or collection.
-public struct Stride<Base: Sequence> {
+public struct StrideSequence<Base: Sequence>: Sequence {
```
Comment (timvermeulen) [inline Sources/Algorithms/Stride.swift:56]: Could you reinstate these doc comments (without the "or collection" bit)?

File: Sources/Algorithms/Stride.swift [hunk 3]
```diff
@@ -88,12 +102,36 @@ extension Stride: Sequence {
...
   
   @inlinable
-  public func makeIterator() -> Stride<Base>.Iterator {
+  public func makeIterator() -> Iterator {
     Iterator(iterator: base.makeIterator(), stride: stride)
   }
 }
 
-extension Stride: Collection where Base: Collection {
+/// A wrapper that strides over a base sequence or collection.
```
Comment (timvermeulen) [inline Sources/Algorithms/Stride.swift:110]:
```suggestion
/// A wrapper that strides over a base collection.
```

**Comment by timvermeulen** [comment]
Thanks @ollieatkinson, this looks great! Awesome job on including some measurements that show this indeed solves the issue we were seeing. Let's rename the `Stride` type to `StrideCollection` for more consistency with the `StrideSequence` name, and then I think we can land this.

**Pull request merged**